Sources/ManageLanguages.php

```diff
@@ -1233,21 +1401,66 @@ function ModifyLanguage()
 		{
 			checkSession();
 
-			$file_contents = implode('', file($current_file));
+			$file_contents = file_get_contents($current_file);
+
 			foreach ($final_saves as $save)
-				$file_contents = strtr($file_contents, array($save['find'] => $save['replace']));
+			{
+				if (!empty($save['is_regex']))
+					$file_contents = preg_replace('~' . $save['find'] . '~' . ($context['utf8'] ? 'u' : ''), $save['replace'], $file_contents);
+				else
+					$file_contents = str_replace($save['find'], $save['replace'], $file_contents);
+			}
 
 			// Save the actual changes.
-			$fp = fopen($current_file, 'w+');
-			fwrite($fp, strtr($file_contents, array("\r" => '')));
-			fclose($fp);
+			file_put_contents($current_file, $file_contents);
 
 			$madeSave = true;
 		}
 
 		// Another restore.
 		$txt = $old_txt;
 
+		// If they can add new language entries, make sure the UI is set up for that
+		if (!empty($context['can_add_lang_entry']))
+		{
+			// Make sure the Add button has a place to show up.
+			foreach ($context['can_add_lang_entry'] as $group => $value)
+			{
+				if (!isset($context['file_entries'][$group]))
+					$context['file_entries'][$group] = array();
+			}
+
+			addInlineJavaScript('
+				function add_lang_entry(group) {
+					var key = prompt("' . $txt['languages_enter_key'] . '");
+
+					if (key !== null) {
+						++entry_num;
+
+						var array_regex = /^(.*)(\[[^\[\]]*\])$/
+						var result = array_regex.exec(key);
+						if (result != null) {
+							key = result[1];
+							var subkey = result[2];
+						} else {
+							var subkey = "";
+						}
+
+						var bracket_regex = /[\[\]]/
+						if (bracket_regex.test(key)) {
+							alert("' . $txt['languages_invalid_key'] . '" + key + subkey);
+							return;
+						}
+
+						$("#language_" + group).append("<dt><span>" + key + subkey + "</span></dt> <dd id=\"entry_" + entry_num + "\"><input id=\"entry_" + entry_num + "_edit\" class=\"entry_toggle\" type=\"checkbox\" name=\"edit[" + key + "]" + subkey + "\" value=\"add\" data-target=\"#entry_" + entry_num + "\" checked> <label for=\"entry_" + entry_num + "_edit\">' . $txt['edit'] . '</label> <input type=\"hidden\" class=\"entry_oldvalue\" name=\"grp[" + key + "]\" value=\"" + group + "\"> <textarea name=\"entry[" + key + "]" + subkey + "\" class=\"entry_textfield\" cols=\"40\" rows=\"1\" style=\"width: 96%; margin-bottom: 2em;\"></textarea></dd>");
+					}
+				};');
+
+			addInlineJavaScript('
```

Comment:
Is ther a reason why do two addinlinejs calls instead of one?

Reply:
No, there is not.

Reply:
Oh, yes there is. One is deferred and the other is not.